presentation/components/src/presentation-components/propertygrid/DataProvider.ts

```diff
@@ -632,45 +428,114 @@ class PropertyDataBuilder {
 
-    // put everything in return format
-    const result: CategorizedRecords = {
-      categories: propertyCategories
-        .filter(({ categoryHasParent }) => !categoryHasParent)
-        .map(({ category }) => category),
-      records: {},
+  private buildFavoriteFieldAncestors(field: Field) {
+    let parentField = field.parent;
+    while (parentField) {
+      parentField = parentField.clone();
+      parentField.nestedFields = [field];
+      parentField.category = this._categoriesCache.getFavoriteCategory(parentField.category);
+      field = parentField;
+      parentField = parentField.parent;
+    }
+    field.rebuildParentship();
+  }
+  private createFavoriteFieldsHierarchy(hierarchy: FieldHierarchy): FieldHierarchy {
+    const favoriteField = hierarchy.field.clone();
+    favoriteField.category = this._categoriesCache.getFavoriteCategory(hierarchy.field.category);
+    this.buildFavoriteFieldAncestors(favoriteField);
+    return {
+      field: favoriteField,
+      childFields: hierarchy.childFields.map((c) => this.createFavoriteFieldsHierarchy(c)),
     };
-    categorizedRecords.forEach((recs, categoryName) => {
-      if (recs.length)
-        result.records[categoryName] = recs.map((r) => r.record);
-    });
-    return result;
+  }
+  private createFavoriteFieldsList(fieldHierarchies: FieldHierarchy[]): FieldHierarchy[] {
+    const favorites: FieldHierarchy[] = [];
+    fieldHierarchies.forEach((fh) => traverseFieldHierarchy(fh, (hierarchy) => {
+      if (!this._props.callbacks.isFavorite(hierarchy.field))
+        return true;
+
+      addFieldHierarchy(favorites, this.createFavoriteFieldsHierarchy(hierarchy));
+      return false;
+    }));
+    return favorites;
+  }
+  public processFieldHierarchies(props: ProcessFieldHierarchiesProps): void {
+    super.processFieldHierarchies(props);
+    this._favoriteFieldHierarchies = this.createFavoriteFieldsList(props.hierarchies);
+    props.hierarchies.push(...this._favoriteFieldHierarchies);
+  }
+
+  public startCategory(props: StartCategoryProps): boolean {
+    this._categoriesStack.push(props.category);
+    return true;
+  }
+  public finishCategory(): void { this._categoriesStack.pop(); }
+
+  public startStruct(props: StartStructProps): boolean {
+    if (this.shouldSkipField(props.hierarchy.field, () => !Object.keys(props.rawValues).length))
+      return false;
+
+    return super.startStruct(props);
+  }
+
+  public startArray(props: StartArrayProps): boolean {
+    if (this.shouldSkipField(props.hierarchy.field, () => !props.rawValues.length))
+      return false;
+
+    return super.startArray(props);
+  }
+
+  public processPrimitiveValue(props: ProcessPrimitiveValueProps): void {
+    if (this.shouldSkipField(props.field, () => (null === props.rawValue || undefined === props.rawValue || "" === props.rawValue)))
+      return;
+
+    super.processPrimitiveValue(props);
+  }
+
+  private shouldSkipField(field: Field, isValueEmpty: () => boolean): boolean {
```

Comment:
```suggestion
  private shouldSkipField(field: Field, isValueEmpty: boolean): boolean {
```

Reply:
I intentionally made it a function to avoid having to evaluate whether the value is empty for cases when we determine that the field should be skipped earlier.

Reply:
Ok, thanks.